src/main/java/com/amazon/ion/impl/IonReaderContinuableCoreBinary.java

```diff
@@ -1553,51 +1560,78 @@ private void readGroupExpression(Macro.ParameterEncoding encoding, List<Expressi
             }
             int startIndex = expressions.size();
             expressions.add(Expression.Placeholder.INSTANCE);
+            boolean isSingleton = true;
             while (nextGroupedValue() != Event.NEEDS_INSTRUCTION) {
                 readValueAsExpression(expressions);
+                isSingleton = false;
+            }
+            if (requireSingleton && !isSingleton) {
+                throw new IonException(String.format(
+                    "Parameter %s with cardinality %s must not contain multiple expressions.",
+                    parameter.getVariableName(),
+                    parameter.getCardinality().name())
+                );
             }
             if (exitArgumentGroup() == Event.NEEDS_DATA) {
                 throw new UnsupportedOperationException("TODO: support continuable parsing of macro arguments.");
             }
             expressions.set(startIndex, new Expression.ExpressionGroup(startIndex, expressions.size()));
         }
 
+        /**
+         * Adds an expression that conveys that the parameter was not present (void).
+         * @param expressions receives the expressions as they are materialized.
+         */
+        private void addVoidExpression(List<Expression.EExpressionBodyExpression> expressions) {
+            int startIndex = expressions.size();
+            expressions.add(new Expression.ExpressionGroup(startIndex, startIndex + 1));
+        }
+
         /**
          * Reads a single parameter to a macro invocation.
          * @param parameter information about the parameter from the macro signature.
          * @param parameterPresence the presence bits dedicated to this parameter.
          * @param expressions receives the expressions as they are materialized.
          */
         private void readParameter(Macro.Parameter parameter, long parameterPresence, List<Expression.EExpressionBodyExpression> expressions) {
-            Macro.ParameterEncoding encoding = parameter.getType();
             switch (parameter.getCardinality()) {
```

Comment:
Follow up for later. I think it will be more readable (and have less branching) if you flip the order that these two enums are checked. E.g.:
```java
switch (parameterPresence) {
    case PresenceBitmap.EXPRESSION:
        // No need to check the cardinality because a single expression is always valid.
        readSingleExpression(parameter, expressions);
        break;
    case PresenceBitmap.VOID:
        // For the happy case, this will always be false, so the JIT can probably make branch prediction optimizations.
        if (!parameter.getCardinality().canBeVoid) {
            throw new IonException(String.format("Invalid void argument for non-voidable parameter: %s", parameter));
        }
        addVoidExpression(expressions);
        break;
    case PresenceBitmap.GROUP:
        // Note, however, that `parameter` is already being passed as an argument, 
        // so `!parameter.getCardinality().canBeMulti` is redundant.
        readGroupExpression(parameter, expressions, !parameter.getCardinality().canBeMulti);
        break;
    default:
        throw new IllegalStateException("Unreachable: presence bitmap validated but reserved bits found.");
}
```